src/herder/HerderSCPDriver.cpp

```diff
@@ -706,6 +711,12 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
                 {
                     highestTxSet = cTxSet;
                     highest = sv.txSetHash;
+                    if (preserveTxSetCloseTimeAffinity)
```

Comment:
I think we can move this out of the loop and use iterators like I suggested before. It avoids copies, and I think the code is easier to read:
```c++
diff --git a/src/herder/HerderSCPDriver.cpp b/src/herder/HerderSCPDriver.cpp
index 2991f200..b0cc2d08 100644
--- a/src/herder/HerderSCPDriver.cpp
+++ b/src/herder/HerderSCPDriver.cpp
@@ -694,50 +694,43 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
     }

     // take the txSet with the biggest size, highest xored hash that we have
-    TxSetFramePtr bestTxSet;
     auto const preserveTxSetCloseTimeAffinity =
         curProtocolPreservesTxSetCloseTimeAffinity();
+    auto highest = candidateValues.end();
     {
-        Hash highest;
         TxSetFramePtr highestTxSet;
-        for (auto const& sv : candidateValues)
+        for (auto it = candidateValues.begin(); it != candidateValues.end();
+             ++it)
         {
+            auto const& sv = *it;
             TxSetFramePtr cTxSet = mPendingEnvelopes.getTxSet(sv.txSetHash);

             if (cTxSet && cTxSet->previousLedgerHash() == lcl.hash)
             {
-                if (compareTxSets(highestTxSet, cTxSet, highest, sv.txSetHash,
-                                  lcl.header, candidatesHash))
+                if (highest == candidateValues.end() ||
+                    compareTxSets(highestTxSet, cTxSet, highest->txSetHash,
+                                  sv.txSetHash, lcl.header, candidatesHash))
                 {
                     highestTxSet = cTxSet;
-                    highest = sv.txSetHash;
-                    if (preserveTxSetCloseTimeAffinity)
-                    {
-                        comp.closeTime = sv.closeTime;
-                        comp.ext.v(STELLAR_VALUE_SIGNED);
-                        comp.ext.lcValueSignature() = sv.ext.lcValueSignature();
-                    }
+                    highest = it;
                 }
             }
         }
-        // make a copy as we're about to modify it and we don't want to mess
-        // with the txSet cache
-        bestTxSet = std::make_shared<TxSetFrame>(*highestTxSet);
     }

+    comp = *highest;
     if (!preserveTxSetCloseTimeAffinity)
     {
         comp.closeTime = maxCloseTime;
         comp.ext.v(STELLAR_VALUE_BASIC);
     }
+    comp.upgrades.clear();
     for (auto const& upgrade : upgrades)
     {
         Value v(xdr::xdr_to_opaque(upgrade.second));
         comp.upgrades.emplace_back(v.begin(), v.end());
     }

-    comp.txSetHash = bestTxSet->getContentsHash();
-
     auto res = wrapStellarValue(comp);
     return res;
 }
```

Reply:
Okay, I like that too.  Adopted.

Reply:
In fact, as long as we're using iterators, why not go all the way and use the STL?  No explicit copying in the loop, and indeed no explicit loop, just a functional expression that we're finding the maximum of the list with an inline comparison function.

```
    // take the txSet with the biggest size, highest xored hash that we have
    using HighestPair = std::pair<StellarValue, TxSetFramePtr>;
    HighestPair highestPair = std::accumulate(
        candidateValues.begin(), candidateValues.end(), HighestPair(),
        [&](HighestPair const& hp, StellarValue const& sv) {
            auto const& highestTxSet = hp.second;
            auto const& svSetHash = sv.txSetHash;
            auto cTxSet = mPendingEnvelopes.getTxSet(svSetHash);
            return (cTxSet && cTxSet->previousLedgerHash() == lcl.hash &&
                    (!highestTxSet ||
                     compareTxSets(highestTxSet, cTxSet, hp.first.txSetHash,
                                   svSetHash, lcl.header, candidatesHash)))
                       ? HighestPair(sv, cTxSet)
                       : hp;
        });

    comp = highestPair.first;
    if (!curProtocolPreservesTxSetCloseTimeAffinity())
    {
        comp.closeTime = maxCloseTime;
        comp.ext.v(STELLAR_VALUE_BASIC);
    }
    comp.upgrades.clear();
```